deltachat-contact-tools/src/lib.rs

```diff
@@ -29,10 +29,116 @@ use std::fmt;
+pub struct VcardContact {
+    /// The email address, vcard property `email`
+    pub addr: String,
+    /// The contact's display name, vcard property `fn`
+    pub display_name: String,
+    /// The contact's public PGP key, vcard property `key`
+    pub key: Option<String>,
+    /// The contact's profile photo (=avatar), vcard property `photo`
+    pub profile_photo: Option<String>,
+    /// The timestamp when the vcard was created / last updated, vcard property `rev`
+    pub timestamp: Result<u64>,
+}
+
+pub fn parse_vcard(vcard: String) -> Result<Vec<VcardContact>> {
+    fn remove_prefix<'a>(s: &'a str, prefix: &str) -> Option<&'a str> {
+        let start_of_s = s.get(..prefix.len())?;
+
+        if start_of_s.eq_ignore_ascii_case(prefix) {
+            s.get(prefix.len()..)
+        } else {
+            None
+        }
+    }
+    fn vcard_property<'a>(s: &'a str, property: &str) -> Option<&'a str> {
+        let remainder = remove_prefix(s, property)?;
+
+        // TODO this doesn't handle the case where there are quotes around a colon
+        let (_params, value) = remainder.split_once(':')?;
+        Some(value)
+    }
+    fn parse_datetime(datetime: Option<&str>) -> Result<u64> {
+        let datetime = datetime.context("No timestamp in vcard")?;
+
+        // According to https://www.rfc-editor.org/rfc/rfc6350#section-4.3.5, the timestamp
+        // is in ISO.8601.2004 format. DateTime::parse_from_rfc3339() apparently parses
+        // ISO.8601, but fails to parse any of the examples given.
+        // So, instead just parse using a format string.
+        let datetime =
+            DateTime::parse_from_str(datetime, "%Y%m%dT%H%M%S%#z") // Parses 19961022T140000Z, 19961022T140000-05, or 19961022T140000-0500
+                .or_else(|_| DateTime::parse_from_str(datetime, "%Y%m%dT%H%M%S"))?; // Parses 19961022T140000
+        let timestamp = datetime.timestamp().try_into()?;
+        Ok(timestamp)
+    }
+
+    let mut lines = vcard.lines().peekable();
+    let mut contacts = Vec::new();
+
+    while lines.peek().is_some() {
+        // Skip to the start of the vcard:
+        for line in lines.by_ref() {
+            if line.eq_ignore_ascii_case("BEGIN:VCARD") {
+                break;
+            }
+        }
+
+        let mut display_name = "";
+        let mut addr = "";
+        let mut key = None;
+        let mut photo = None;
+        let mut datetime = None;
+
+        for line in lines.by_ref() {
+            if let Some(email) = vcard_property(line, "email") {
+                addr = email;
+            } else if let Some(name) = vcard_property(line, "fn") {
+                display_name = name;
+            } else if let Some(k) = remove_prefix(line, "KEY;PGP;ENCODING=BASE64:")
+                .or_else(|| remove_prefix(line, "KEY;TYPE=PGP;ENCODING=b:"))
+                .or_else(|| remove_prefix(line, "KEY:data:application/pgp-keys;base64,"))
+            {
+                key = Some(key.unwrap_or(k));
```

Comment:
Why not just `Some(k)`? For `addr` and `display_name` you overwrite the previous value

Reply:
> For `addr` and `display_name` you overwrite the previous value

I didn't mean to, because it seemed most sensible to use the first value if there are multiple - I think we shouldn't overwrite `addr` and `display_name`, either. I'll fix that if you don't disagree

Reply:
So, i also used the first value for `addr` and `display_name`, even if it's empty. This way the code is the same for all properties. I think it shouldn't be empty anyway